src/sparse/impl/KokkosSparse_spgemm_jacobi_sparseacc_impl.hpp

```diff
@@ -0,0 +1,1534 @@
+	    for (nnz_lno_t trial = hash; trial < team_cuckoo_key_size; ){
+	      if (keys[trial] == my_b_col){
+		Kokkos::atomic_add(vals + trial, my_b_val);
+		fail = 0;
+		break;
+	      }
+	      else if (keys[trial] == init_value){
+		if (Kokkos::atomic_compare_exchange_strong(keys + trial, init_value, my_b_col)){
+		  Kokkos::atomic_add(vals + trial, my_b_val);
+		  fail = 0;
+		  break;
+		}
+	      }
+	      else {
+		++trial;
+	      }
+	    }
+
+	    if (fail){
+	      for (nnz_lno_t trial = 0; trial < hash; ){
+		if (keys[trial] == my_b_col){
+		  Kokkos::atomic_add(vals + trial, my_b_val);
+		  fail = 0;
+		  break;
+		}
+		else if (keys[trial] == init_value){
+		  if (Kokkos::atomic_compare_exchange_strong(keys + trial, init_value, my_b_col)){
+		    Kokkos::atomic_add(vals + trial, my_b_val);
+		    fail = 0;
+		    break;
+		  }
+		}
+		else {
+		  ++trial;
+		}
+	      }
+	    } // end if (fail)
+	  } // end for (nnz_lno_t vector_read_shift = vector_shift ...
+
+
+	  teamMember.team_barrier();
+	  for (nnz_lno_t  my_index = vector_shift; my_index < team_cuckoo_key_size;my_index += bs){
+	    nnz_lno_t my_key = keys[my_index];
+	    if (my_key != init_value){
+	      scalar_t my_val = vals[my_index];
+	      nnz_lno_t write_index = Kokkos::atomic_fetch_add(used_hash_sizes, atomic_incr_type(1));
+	      c_row[write_index] = my_key;
+	      c_row_vals[write_index] = my_val;
+	    }
+	  }
+	} // end for (nnz_lno_t  my_index = vector_shift ...
+      }
+
+      // The thread-flat-parallel implementation for the case one row does not fit into shmem
+      KOKKOS_INLINE_FUNCTION
+      void operator()(const GPUTag6&, const team_member_t & teamMember) const {
+
+	const nnz_lno_t init_value = -1;
+	nnz_lno_t team_row_begin = teamMember.league_rank()  * team_work_size;
+	const nnz_lno_t team_row_end = KOKKOSKERNELS_MACRO_MIN(team_row_begin + team_work_size, numrows);
+
+	char *all_shared_memory = (char *) (teamMember.team_shmem().get_shmem(shared_memory_size));
+	volatile nnz_lno_t *used_hash_sizes = (volatile nnz_lno_t *) (all_shared_memory);
+	all_shared_memory += sizeof(nnz_lno_t) * 2;
+
+	nnz_lno_t *keys = (nnz_lno_t *) (all_shared_memory);
+	all_shared_memory += sizeof(nnz_lno_t) * team_cuckoo_key_size;
+	scalar_t *vals = KokkosKernels::Impl::alignPtr<char*, scalar_t>(all_shared_memory);
+
+	int thread_rank =  teamMember.team_rank();
+	int vector_rank = 0;
+	typedef typename std::remove_reference<decltype(*used_hash_sizes)>::type atomic_incr_type;
+	Kokkos::parallel_scan(Kokkos::ThreadVectorRange(teamMember, vector_size), [&] (const int /* threadid */, int &update, const bool final) {
+	    if (final){
+	      vector_rank = update;
+	    }
+	    update += 1;
+	  });
+
+	int bs = vector_size * suggested_team_size;
+	int vector_shift = thread_rank * vector_size + vector_rank;
+
+	for (nnz_lno_t row_index = team_row_begin; row_index < team_row_end; ++row_index){
+
+	  teamMember.team_barrier();
+
+	  scalar_t mult = -omega*dinv(row_index,0);
```

Comment:
In this case, it's a whole team accessing the same address for dinv so it's not quite optimal in bandwidth terms, but I doubt there's a win from copying to shared first.